samples/s02-instances/hostCode.cpp

```diff
@@ -0,0 +1,342 @@
+    { "vertex", GPRT_BUFFER, GPRT_OFFSETOF(TrianglesGeomData,vertex)},
+    { /* sentinel to mark end of list */ }
+  };
+  GPRTGeomType trianglesGeomType
+    = gprtGeomTypeCreate(context,
+                        GPRT_TRIANGLES,
+                        sizeof(TrianglesGeomData),
+                        trianglesGeomVars,-1);
+  gprtGeomTypeSetClosestHitProg(trianglesGeomType,0,
+                           module,"TriangleMesh");
+
+  // -------------------------------------------------------
+  // set up miss prog
+  // -------------------------------------------------------
+  GPRTVarDecl missProgVars[]
+    = {
+    { "color0", GPRT_FLOAT3, GPRT_OFFSETOF(MissProgData,color0)},
+    { "color1", GPRT_FLOAT3, GPRT_OFFSETOF(MissProgData,color1)},
+    { /* sentinel to mark end of list */ }
+  };
+  GPRTMiss miss
+    = gprtMissCreate(context,module,"miss",sizeof(MissProgData),
+                        missProgVars,-1);
+
+  // -------------------------------------------------------
+  // set up ray gen program
+  // -------------------------------------------------------
+  GPRTVarDecl rayGenVars[] = {
+    { "fbSize",        GPRT_INT2,   GPRT_OFFSETOF(RayGenData,fbSize)},
+    { "fbPtr",         GPRT_BUFFER, GPRT_OFFSETOF(RayGenData,fbPtr)},
+    { "world",         GPRT_ACCEL,  GPRT_OFFSETOF(RayGenData,world)},
+    { "camera.pos",    GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.pos)},
+    { "camera.dir_00", GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.dir_00)},
+    { "camera.dir_du", GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.dir_du)},
+    { "camera.dir_dv", GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.dir_dv)},
+    { /* sentinel to mark end of list */ }
+  };
+
+  GPRTRayGen rayGen
+    = gprtRayGenCreate(context,module,"simpleRayGen",
+                      sizeof(RayGenData),
+                      rayGenVars,-1);
+
+  // ##################################################################
+  // set the parameters for those kernels
+  // ##################################################################
+
+  // Setup pixel frame buffer
+  GPRTBuffer frameBuffer
+    = gprtDeviceBufferCreate(context,GPRT_INT,fbSize.x*fbSize.y);
+  gprtRayGenSetBuffer(rayGen,"fbPtr", frameBuffer);
+  gprtRayGenSet2iv(rayGen,"fbSize", (int32_t*)&fbSize);
+
+  // Miss program checkerboard background colors
+  gprtMissSet3f(miss,"color0",0.1f,0.1f,0.1f);
+  gprtMissSet3f(miss,"color1",.0f,.0f,.0f);
+
+  LOG("building geometries ...");
+
+  // Create our cube mesh
+  GPRTBuffer vertexBuffer
+    = gprtDeviceBufferCreate(context,GPRT_FLOAT3,NUM_VERTICES,vertices);
+  GPRTBuffer indexBuffer
+    = gprtDeviceBufferCreate(context,GPRT_INT3,NUM_INDICES,indices);  
+  GPRTGeom trianglesGeom
+    = gprtGeomCreate(context,trianglesGeomType);
+  gprtTrianglesSetVertices(trianglesGeom,vertexBuffer,
+                           NUM_VERTICES,sizeof(float3),0);
+  gprtTrianglesSetIndices(trianglesGeom,indexBuffer,
+                          NUM_INDICES,sizeof(int3),0);
```

Comment:
A couple of thoughts here:

  - Are there times when we might pass an array that's non-float to `gprtTriangleSetVertices`? In other words, should the `stride` argument be in number of floats rather than bytes?
  - Would it be reasonable to make the `offset` argument default to `0`?

```suggestion
  gprtTrianglesSetVertices(trianglesGeom,
                                               vertexBuffer,
                                               NUM_VERTICES,
                                               sizeof(float3),
                                               0);
  gprtTrianglesSetIndices(trianglesGeom,
                                              indexBuffer,
                                              NUM_INDICES,
                                              sizeof(int3),
                                              0);
```

Reply:
For your first question,, floats are required. RT cores dictate these must be floats. It is a bit odd we use bytes here for stride, but then count for buffers. 

I do this to match our old library OWL, so that porting is easier. 

For your second question, we could be defaulting to 0 for the offset here yeah. We honestly could also probably default to sizeof(float3) for the stride.

Reply:
Yeah that’s what I’m thinking. If that type is always the same why not get rid of the ‘size of’ for these calls.

Reply:
The big argument against this is that it will prevent users like @szellmann from easily porting over OWL code to GPRT.

Reply:
The defaults is a good idea though. Just added them now. Should be resolved.

Reply:
I guess it depends on what our priorities are. If we want to stay more aligned with OWL, we'd leave it, but, outside of that perspective, this seems like a clear improvement to the API to me. I'm curious if there's a reason OWL had the signatures setup this way. If we asked the OWL devs do you think they would have designed it differently in hindsight?

I'm thinking we could resolve small differences in the API with  a section in the documentation called "Porting OWL code"or something like that.

At any rate, I'll leave it up to what you think is best! Just trying to clarify goals a bit 😃